qemu/tests/virtio_driver_sign_check.py

```diff
@@ -1,87 +1,121 @@
 
     :param test: QEMU test object
     :param params: Dictionary with the test parameters
     :param env: Dictionary with test environment.
     """
     vm = env.get_vm(params["main_vm"])
     vm.verify_alive()
-
-    def is_sdksetup_finished():
-        s, o = session.cmd_status_output("tasklist | find \"SDK\"")
-        if s:
-            return True
-        else:
-            return False
-
     timeout = float(params.get("login_timeout", 240))
     session = vm.wait_for_login(timeout=timeout)
-    signtool_install = params.get("signtool_install")
-    au3_link = params.get("winsdk_au3", "autoit/winsdk.au3")
-    au3_link = os.path.join(test.bindir, au3_link)
     signtool_cmd = params.get("signtool_cmd")
     list_files_cmd = params.get("list_files_cmd")
-    drive_list = params.get("drive_list")
-    vm.copy_files_to(au3_link, "c:\\")
-    drivers = {}
-    logging.info("Install sdk in guest if signtool is not available.")
-    session.cmd_status_output(signtool_install, timeout=360)
-    # Wait until guest start install sdk.
-    time.sleep(10)
-    logging.info("Waiting for guest sdk setup ...")
-    utils_misc.wait_for(is_sdksetup_finished, timeout=1800)
-    results_all = """All the signature check log:\n"""
+    drive_list = params.objects("drive_list")
+    os_dirs = params.objects("os_dirs")
+    arch_dir = params.get("arch_dir")
     fails_log = """Failed signature check log:\n"""
+    results_all = """All the signature check log:\n"""
     fail_num = 0
     fail_drivers = []
-    logging.info("Running SignTool command in guest...")
+    virtio_drivers = []
+    drivers = {}
+    need_test_driver = params.objects("tested_drivers")
+    get_disk_letter = params.get("get_disk_letter_cmd")
+
+    def cat_filter(driver):
+        pattern = r'(\w:)\\(.+?)\\(.+?)\\(.+?)[\.|\\]'
+        ret = re.match(pattern, driver, re.I)
+        disk_name = ret.group(1)
+        os_name = ret.group(3)
+
+        if disk_name == "A:":
+            sub_os_name = os_dirs[0]
+            plat_name = ret.group(2)
+            driver_dir = ret.group(4)
+        else:
+            sub_os_name = os_dirs[1]
+            driver_dir = ret.group(2)
+            plat_name = ret.group(4)
+
+        if sub_os_name in os_name:
+            if arch_dir in plat_name:
+                if driver_dir in need_test_driver:
+                    return True
+
+    def match_cat(cat_element):
+        pattern = r'\w:\\(.+?)\\(.+?)\\(.+?)[\.|\\]'
+        match_ret = re.match(pattern, cat_element)
+        driver_dir = match_ret.group()
+
+        driver_sys = filter(lambda x: driver_dir in x, drivers['.sys'])
+        driver_inf = filter(lambda x: driver_dir in x, drivers['.inf'])
+        driver_wdf = filter(lambda x: driver_dir in x, drivers['Wdf'])
+
+        cat_sys = map(lambda x: (cat_element, x), driver_sys)
+        cat_inf = map(lambda x: (cat_element, x), driver_inf)
+        cat_wdf = map(lambda x: (cat_element, x), driver_wdf)
```

Comment:
maybe these codes above can be written in a loop

Reply:
Hi wyu, 
Here, it is not suitable use a loop.  Because it doesn't need to judge conditions, so while loop  is not suitable;  and if use "for loop", the loop format like "for tmp in list/string:", the tmp should be a string or character, but in our code, the drivers['sys']  is a list, and the drivers is a dict, they are not suitable to use "for loop" .